nominal/_utils/multipart_downloader.py

```diff
@@ -0,0 +1,347 @@
+                failed[dest] = ex
+                logger.error("Failed part for %s @%d: %s", dest, start, ex, exc_info=ex)
+
+        # Only count as success if it planned and never failed
+        ok = [d for d in plans.keys() if d in succeeded and d not in failed]
+        return DownloadResults(succeeded=ok, failed=failed)
+
+    # ---- planning helpers ----
+
+    def _make_session(self, pool_size: int) -> requests.Session:
+        retries = Retry(
+            total=5,
+            backoff_factor=0.5,
+            status_forcelist=(429, 500, 502, 503, 504),
+            allowed_methods=frozenset(["GET", "HEAD"]),
+        )
+        s = requests.Session()
+        adapter = HTTPAdapter(max_retries=retries, pool_maxsize=pool_size)
+        s.mount("https://", adapter)
+        s.mount("http://", adapter)
+        return s
+
+    def _head_or_probe(self, provider: PresignedURLProvider) -> Tuple[int, str | None]:
+        """Discover (total_size, etag). Refresh once if the current URL is stale."""
+        for attempt in range(2):
+            url = provider.get(force=(attempt > 0))
+
+            r = self._session.head(url, timeout=self.timeout)
+            if r.ok and "Content-Length" in r.headers:
+                return int(r.headers["Content-Length"]), r.headers.get("ETag")
+
+            r = self._session.get(url, headers={"Range": "bytes=0-0"}, timeout=self.timeout)
+            if r.ok:
+                total = (
+                    int(r.headers["Content-Range"].split("/")[-1])
+                    if "Content-Range" in r.headers
+                    else int(r.headers["Content-Length"])
+                )
+                return total, r.headers.get("ETag")
+
+            if self._is_expired_status(r):
+                provider.invalidate()
+                continue
+
+            r.raise_for_status()
+
+        raise RuntimeError("Could not determine object size/ETag (presigned URL kept failing)")
+
+    def _iter_ranges(self, total_size: int, part_size: int) -> Iterable[Tuple[int, int, int]]:
+        """Yield (index, start, end) inclusive ranges."""
+        parts = max(1, math.ceil(total_size / part_size))
+        for i in range(parts):
+            start = i * part_size
+            end = min(total_size - 1, start + part_size - 1)
+            yield (i, start, end)
+
+    # ---- IO helpers ----
+
+    def _prepare_destination(self, path: pathlib.Path, *, force: bool) -> None:
+        logger.info("Preparing file destination %s (force=%s)", path, force)
+
+        parent = path.parent
+        if not parent.exists():
+            if force:
+                parent.mkdir(parents=True, exist_ok=True)
+            else:
+                raise FileNotFoundError(f"Output directory does not exist and force=False: {parent}")
+
+        if path.exists():
+            if path.is_dir():
+                raise ValueError(f"Destination exists as a directory: {path}")
+            if force:
+                path.unlink()
```

Comment:
I don't think this is necessary, if there's an error it's possible we delete the files without writing any. `open(p, "w")` will delete the contents, so as long as we throw if the file exists before opening, then we can just open and write